realm/realm-library/src/main/java/io/realm/RealmCache.java

```diff
@@ -492,20 +494,31 @@ synchronized void invokeWithLock(Callback0 callback) {
      * @param configuration configuration object for Realm instance.
      * @throws RealmFileException if copying the file fails.
      */
-    private static void copyAssetFileIfNeeded(RealmConfiguration configuration) {
-        if (configuration.hasAssetFile()) {
-            File realmFile = new File(configuration.getRealmDirectory(), configuration.getRealmFileName());
-
-            copyFileIfNeeded(configuration.getAssetFilePath(), realmFile);
-        }
+    private static void copyAssetFileIfNeeded(final RealmConfiguration configuration) {
+        final File realmFileFromAsset = configuration.hasAssetFile() ?
+                new File(configuration.getRealmDirectory(), configuration.getRealmFileName())
+                : null;
+        final String syncServerCertificateAssetName = ObjectServerFacade.getFacade(
+                configuration.isSyncConfiguration()).getSyncServerCertificateAssetName(configuration);
+
+        if (realmFileFromAsset!= null || !Util.isEmptyString(syncServerCertificateAssetName)) {
+            OsObjectStore.callWithLock(configuration, new Runnable() {
+                @Override
+                public void run() {
+                    if (realmFileFromAsset != null) {
+                        copyFileIfNeeded(configuration.getAssetFilePath(), realmFileFromAsset);
+                    }
 
-        // Copy Sync Server certificate path if available
-        String syncServerCertificateAssetName = ObjectServerFacade.getFacade(configuration.isSyncConfiguration()).getSyncServerCertificateAssetName(configuration);
-        if (!Util.isEmptyString(syncServerCertificateAssetName)) {
-            String syncServerCertificateFilePath = ObjectServerFacade.getFacade(configuration.isSyncConfiguration()).getSyncServerCertificateFilePath(configuration);
+                    // Copy Sync Server certificate path if available
+                    if (!Util.isEmptyString(syncServerCertificateAssetName)) {
```

Comment:
use a final `boolean` in  https://github.com/realm/realm-java/pull/5417/files#diff-79e6ca3beee0a83dd76772285b606694R504 to the redundant check again